### PR TITLE
SCP 2364 Add time tracing for profiling

### PR DESCRIPTION
**Pull request by thealmarty**
- Added `EmitterOption` which is either `NoEmit`, `Emit`, or `EmitWithTimestamp` to specify whether logging should be turned on, and if so, whether it should include time stamp.
- `evaluateCek` and its unsafe version now takes a `Bool` which indicates whether time tracing should be turned on.
- Closes #3577 
- Closes #3583 

Pre-submit checklist:
- Branch
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

**Comment by thealmarty** [comment]
I ran `cabal bench benchmark:nofib` on my branch and compared it with master. See the results [here](https://gist.github.com/thealmarty/1baab37c6a3238a3d854ee8098c22750). It doesn't seem like my changes make things slower.

**Comment by thealmarty** [comment]
I'm going to be away for the next few days. Please feel free to push changes to here directly and squash merge this if you want to start using it.

File: plutus-core/untyped-plutus-core/test/Evaluation/Builtins/Common.hs
Comment (effectfully) [inline plutus-core/untyped-plutus-core/test/Evaluation/Builtins/Common.hs:42]: `unsafeEvaluateCek` is meant for evaluation only, so unless you want to show the logs sometimes, just hardcode that `False` right inside `unsafeEvaluateCek`.

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek.hs
Comment (effectfully) [inline plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek.hs:100]: Replace `Bool` with `EmitterOptiion` here as well please.

Comment (michaelpj) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek.hs:100]: Agree, it's easy enough for consumers to work with. We'll need to export the type too, but that's okay.

Comment (michaelpj) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek.hs:100]: I think just passing it down is simplest.

File: plutus-ledger-api/src/Plutus/V1/Ledger/Api.hs
Comment (michaelpj) [reply plutus-ledger-api/src/Plutus/V1/Ledger/Api.hs:237]: No, I think we do just want to give them a simple toggle for now. Especially in this case, where the emit-with-timestamps mode is never going to be something that you use in the real version (because you'll need to litter your program with traces, which you would never do in reality). So I think it's fine to keep it like this for now. Also generally we want to keep the interface fairly fixed, otherwise we can't change anything without breaking the ledger, which is quite painful.

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs
Comment (michaelpj) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs:392]: This seems very strange. We should never need to show a value of type `ST s UTCTime`, but rather a value of type `UTCTime`.

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs
Comment (michaelpj) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs:651]: Yes, I agree, and this is exactly what the implicit parameters are for.

**Review by michaelpj — APPROVED** [review]
Marty's off for a few days, so I'll push some changes to this.

**Comment by michaelpj** [comment]
I refactored it to make it very like the way budget modes are handled. It doesn't quite go the whole way into having a separate module that defines the modes, with `runCek` just taking a function that produces a mode in `ST`, but we could do that if we wanted to.

**Comment by michaelpj** [comment]
And now in a separate module and matching the budget mode.

**Comment by michaelpj** [comment]
I hate doing this, but I'm in a hurry so here it goes.

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs
Comment (thealmarty) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs]: As the code was, without the instance the log showed "ST actions" instead of the time stamp. With the instance, the log showed the UTC time. It may be fine with the updated code. I'll know when I test it more.

Comment (effectfully) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs]: I mean, in `showsPrec _ x  = showString $ show x` you're calling `show` over the same `x` that `showsPrec` receives. `show x` literally unfolds into `showsPrec 0 x ""`, so does that even terminate?

Or if you don't trust this reasoning, the following also type checks:

```haskell
data NoShow

instance {-# OVERLAPS #-} Show (ST s NoShow) where
    showsPrec _ x  = showString $ show x
``` 

`NoShow` has no `Show` instance and yet the `Show (ST s NoShow)` instance type checks hence it can't defer to the non-existent `Show` instance of `NoShow` and hence the same does not happen for `UTCTime`.

Comment (thealmarty) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs]: It did terminate and showed the time. But Michael is right I should have taken it out of the `ST` and not need this `Show (ST s UTCTime)` instance.

Comment (thealmarty) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs]: It did get stuck in an earlier version but then it did terminate and showed the time. But Michael is right I should have taken it out of the ST and not need this Show (ST s UTCTime) instance.

**Comment by thealmarty** [comment]
Thanks! Very elegant!

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs
Comment (effectfully) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs]: > It did get stuck in an earlier version but then it did terminate and showed the time.

My comment was about the version that is shown above. It does not terminate. Here I went to the trouble of checking that.

Screenshot from 2021-08-04 00-16-21

Comment (thealmarty) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs]: Thanks for letting me know. I ran `cabal test plutus-tx-tests` and looked at the errors and the output log instead. I'm not sure why the time showed up there. Maybe I got confused somehow. Anyway I'll check with what you did here next time too. Also I realized I should avoid writing overlapping custom instance in general! 🤓

**Pull request merged**
